# Lending Club Case Study
## Authors
### Siddakka Saptasagare
### Kumaraguru Muthuraj

## Business Understanding
**You work for a consumer finance company which specialises in lending various types of loans to urban customers. When the company receives a loan application, the company has to make a decision for loan approval based on the applicant’s profile. Two types of risks are associated with the bank’s decision:**
- If the applicant is likely to repay the loan, then not approving the loan results in a loss of business to the company
- If the applicant is not likely to repay the loan, i.e. he/she is likely to default, then approving the loan may lead to a financial loss for the company
<p>
<img src ="https://cdn.upgrad.com/UpGrad/temp/7afbce98-8ecc-41c6-96d8-981cba7d343f/Loan_image.png" alt='Figure 1' style="width:600px;">
<center> <b>Approval - Rejection flow</b> </center> 
 </br>  
</p>

**When a person applies for a loan, there are two types of decisions that could be taken by the company:**

- Loan accepted: If the company approves the loan, there are 3 possible scenarios described below:

> Fully paid: Applicant has fully paid the loan (the principal and the interest rate)

> Current: Applicant is in the process of paying the instalments, i.e. the tenure of the loan is not yet completed. These candidates are not labelled as 'defaulted'.

> Charged-off: Applicant has not paid the instalments in due time for a long period of time, i.e. he/she has defaulted on the loan 

- Loan rejected: The company had rejected the loan (because the candidate does not meet their requirements etc.). Since the loan was rejected, there is no transactional history of those applicants with the company and so this data is not available with the company (and thus in this dataset)

## Business Objectives
**This company is the largest online loan marketplace, facilitating personal loans, business loans, and financing of medical procedures. Borrowers can easily access lower interest rate loans through a fast online interface.**

**The company wants to understand the driving factors (or driver variables) behind loan default, i.e. the variables which are strong indicators of default.  The company can utilise this knowledge for its portfolio and risk assessment.** 

## What are we solving for?
**Lending Club, being headquartered in SFO, CA lends money for cheap interest rates for three major reasons, namely medical, personal and business loans.**

**The history of borrowers between 2007 and 2011 provided will be used to do statistical analysis and provide insights on what parameters of the borrower are related to defaulting. This will help Lending Club avoid lending money to these customers or reduce the loan amount or increase the interest rate.**

## How did we solve?
**We have done extensive EDA on all the parameters in the loan.csv by relating to the meaning in data dictionary. After doing analysis, we have provided multi-perspective analyis and observe the patterns.**

**Once we establish the pattern, we record our conclusion and recommend what the bank should do. Every cell(or group of cells) is dedicated to a set of parameters that we analyze. These indicate the factors impacting defaulting nature of customers.**

**We have used different visualizations to perceive the pattern powerfully which leads to decisions.**

### The insights are provided step-by-step in the form of a story, to graps the most important customer parameters impacting loan defaulting tendency. There are 3 major sections
### A) Extensive EDA.  
### B) EDA and multi-perspective visualizations with insights and conclusions on loan defaulters (charged-off).
### C) EDA experiments supporting insights and conclusions.

# (A) Extensive EDA.  

### Step 1. Load the required libraries, open the loan.csv and cleanup data.

In [329]:
# Loading required python libraries
import math
import re
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import datetime as dt
from datetime import datetime
from dateutil.relativedelta import relativedelta
import warnings
warnings.filterwarnings("ignore")

In [330]:
#Loading and Extracting the loan zip file
#importing the zipfile library required to extract and read the zip file
import zipfile as zf
filename= 'loan.zip'
loan = zf.ZipFile(filename,'r')
# extracting the files using 'extracall' method
print(loan.printdir())
loan.extractall()

File Name                                             Modified             Size
loan.csv                                       2017-05-16 14:10:10     34813575
None


In [331]:
#Creating the dataframe by using the csv file present in the zip file
df = pd.read_csv(loan.open('loan.csv'),low_memory=False, parse_dates=True)
df.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status issue_d  loan_status  \
0           RENT     24000.0            Verified  Dec-11   Fully Paid   
1           RENT     30000.0     Source Verified  Dec-11  Charged Off   
2           RENT     12252.0        Not Verified  Dec-11   Fully Paid   
3           RENT     49200.0     Source Verified  Dec-11   Fully Paid   
4           RENT     80000.0     Source Verified  Dec-11      Current   

  pymnt_plan                                                url  \
0          n  https://lendingclub.com/browse/loanDetail.acti...   
1          n  https://lendingclub.com/browse/loanDetail.acti...   
2          n  https://lendingclub.com/browse/loanDetail.acti...   
3          n  https://lendingclub.com/browse/loanDetail.acti...   
4          n  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state    dti  delinq_2yrs  \
0              Computer    860xx         AZ  27.65            0   
1                  bike    309xx         GA   1.00            0   
2  real estate business    606xx         IL   8.72            0   
3              personel    917xx         CA  20.00            0   
4              Personal    972xx         OR  17.94            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0           Jan-85               1                     NaN   
1           Apr-99               5                     NaN   
2           Nov-01               2                     NaN   
3           Feb-96               1                    35.0   
4           Jan-96               0                    38.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         3        0      13648     83.70%          9   
1                     NaN         3        0       1687      9.40%          4   
2                     NaN         2        0       2956     98.50%         10   
3                     NaN        10        0       5598        21%         37   
4                     NaN        15        0      27783     53.90%         38   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   f       0.00           0.00   5863.155187   
1                   f       0.00           0.00   1008.710000   
2                   f       0.00           0.00   3005.666844   
3                   f       0.00           0.00  12231.890000   
4                   f     524.06         524.06   3513.330000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          5833.84          5000.00         863.16     

In [332]:
#Checking the shape of the dataframe
print(df.shape)

(39717, 111)


In [333]:
#We observed that there are more then 100 columns present in our dataframe
#Lets use set_option to display all the columns of df
pd.set_option('display.max_columns', 200)
df.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status issue_d  loan_status  \
0           RENT     24000.0            Verified  Dec-11   Fully Paid   
1           RENT     30000.0     Source Verified  Dec-11  Charged Off   
2           RENT     12252.0        Not Verified  Dec-11   Fully Paid   
3           RENT     49200.0     Source Verified  Dec-11   Fully Paid   
4           RENT     80000.0     Source Verified  Dec-11      Current   

  pymnt_plan                                                url  \
0          n  https://lendingclub.com/browse/loanDetail.acti...   
1          n  https://lendingclub.com/browse/loanDetail.acti...   
2          n  https://lendingclub.com/browse/loanDetail.acti...   
3          n  https://lendingclub.com/browse/loanDetail.acti...   
4          n  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state    dti  delinq_2yrs  \
0              Computer    860xx         AZ  27.65            0   
1                  bike    309xx         GA   1.00            0   
2  real estate business    606xx         IL   8.72            0   
3              personel    917xx         CA  20.00            0   
4              Personal    972xx         OR  17.94            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0           Jan-85               1                     NaN   
1           Apr-99               5                     NaN   
2           Nov-01               2                     NaN   
3           Feb-96               1                    35.0   
4           Jan-96               0                    38.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         3        0      13648     83.70%          9   
1                     NaN         3        0       1687      9.40%          4   
2                     NaN         2        0       2956     98.50%         10   
3                     NaN        10        0       5598        21%         37   
4                     NaN        15        0      27783     53.90%         38   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   f       0.00           0.00   5863.155187   
1                   f       0.00           0.00   1008.710000   
2                   f       0.00           0.00   3005.666844   
3                   f       0.00           0.00  12231.890000   
4                   f     524.06         524.06   3513.330000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          5833.84          5000.00         863.16     

In [334]:
#Lets drop the columns of the df having all NULL or SAME values.
#If half the rows in a column is empty drop the column.
thresh = len(df) * .5 
df.dropna(thresh = thresh, axis = 1, inplace = True)
print(df.shape)
# Lets reset the row index from default to id
df.set_index(['id'], inplace=True)
df.head()

(39717, 54)


member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
id                                                                        
1077501    1296599       5000         5000           4975.0   36 months   
1077430    1314167       2500         2500           2500.0   60 months   
1077175    1313524       2400         2400           2400.0   36 months   
1076863    1277178      10000        10000          10000.0   36 months   
1075358    1311748       3000         3000           3000.0   60 months   

        int_rate  installment grade sub_grade                 emp_title  \
id                                                                        
1077501   10.65%       162.87     B        B2                       NaN   
1077430   15.27%        59.83     C        C4                     Ryder   
1077175   15.96%        84.33     C        C5                       NaN   
1076863   13.49%       339.31     C        C1       AIR RESOURCES BOARD   
1075358   12.69%        67.79     B        B5  University Medical Group   

        emp_length home_ownership  annual_inc verification_status issue_d  \
id                                                                          
1077501  10+ years           RENT     24000.0            Verified  Dec-11   
1077430   < 1 year           RENT     30000.0     Source Verified  Dec-11   
1077175  10+ years           RENT     12252.0        Not Verified  Dec-11   
1076863  10+ years           RENT     49200.0     Source Verified  Dec-11   
1075358     1 year           RENT     80000.0     Source Verified  Dec-11   

         loan_status pymnt_plan  \
id                                
1077501   Fully Paid          n   
1077430  Charged Off          n   
1077175   Fully Paid          n   
1076863   Fully Paid          n   
1075358      Current          n   

                                                       url  \
id                                                           
1077501  https://lendingclub.com/browse/loanDetail.acti...   
1077430  https://lendingclub.com/browse/loanDetail.acti...   
1077175  https://lendingclub.com/browse/loanDetail.acti...   
1076863  https://lendingclub.com/browse/loanDetail.acti...   
1075358  https://lendingclub.com/browse/loanDetail.acti...   

                                                      desc         purpose  \
id                                                                           
1077501    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1077430    Borrower added on 12/22/11 > I plan to use t...             car   
1077175                                                NaN  small_business   
1076863    Borrower added on 12/21/11 > to pay for prop...           other   
1075358    Borrower added on 12/21/11 > I plan on combi...           other   

                        title zip_code addr_state    dti  delinq_2yrs  \
id                                                                      
1077501              Computer    860xx         AZ  27.65            0   
1077430                  bike    309xx         GA   1.00            0   
1077175  real estate business    606xx         IL   8.72            0   
1076863              personel    917xx         CA  20.00            0   
1075358              Personal    972xx         OR  17.94            0   

        earliest_cr_line  inq_last_6mths  open_acc  pub_rec  revol_bal  \
id                                                                       
1077501           Jan-85               1         3        0      13648   
1077430           Apr-99               5         3        0       1687   
1077175           Nov-01               2         2        0       2956   
1076863           Feb-96               1        10        0       5598   
1075358           Jan-96               0        15        0      27783   

        revol_util  total_acc initial_list_status  out_prncp  out_prncp_inv  \
id                                                                            


In [335]:
# Checking the null values of data frame
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 39717 entries, 1077501 to 87023
Data columns (total 53 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   member_id                   39717 non-null  int64  
 1   loan_amnt                   39717 non-null  int64  
 2   funded_amnt                 39717 non-null  int64  
 3   funded_amnt_inv             39717 non-null  float64
 4   term                        39717 non-null  object 
 5   int_rate                    39717 non-null  object 
 6   installment                 39717 non-null  float64
 7   grade                       39717 non-null  object 
 8   sub_grade                   39717 non-null  object 
 9   emp_title                   37258 non-null  object 
 10  emp_length                  38642 non-null  object 
 11  home_ownership              39717 non-null  object 
 12  annual_inc                  39717 non-null  float64
 13  verification_status      

##### In the following step, we see that all of the columns have their data filled. But if you look at the values visually, we see that they are repeated to the extent where it doesn't give any perspective. Drop these columns! 

In [336]:
#print(df['pymnt_plan'].value_counts())
#print(df['url'].value_counts())

#CAN do mining on 'desc' text, but the reasons are already condensed in the 'purpose' column.
#print(df['desc'].value_counts())
#print(df['initial_list_status'].value_counts())
#print(df['collections_12_mths_ex_med'].value_counts())
#print(df['policy_code'].value_counts())
#print(df['application_type'].value_counts())
#print(df['acc_now_delinq'].value_counts())
#print(df['delinq_amnt'].value_counts())
#print(df['tax_liens'].value_counts())

In [337]:
#Let's drop the irrelevent columns to get proper and structured data frame for analysis
#We are dropping the columns because the values of the column or either "0" or same
df.drop(['pymnt_plan','url','desc','initial_list_status','collections_12_mths_ex_med','policy_code','application_type','acc_now_delinq','chargeoff_within_12_mths','delinq_amnt','tax_liens'], inplace=True, axis=1)
print(df.shape)

(39717, 42)


### Standardization and data formating of columns.
#### Add columns here as and when required.

In [338]:
# Lets replace special character and convert interest rate and revol_util to float
df.int_rate = df['int_rate'].str.strip('%').astype('float')
df.revol_util = df['revol_util'].str.strip('%').astype('float').round()

getTerm = lambda x: x.strip().split(' ')[0]
df['term'] = (df['term'].apply(getTerm)).astype('int')

df.head()

member_id  loan_amnt  funded_amnt  funded_amnt_inv  term  int_rate  \
id                                                                            
1077501    1296599       5000         5000           4975.0    36     10.65   
1077430    1314167       2500         2500           2500.0    60     15.27   
1077175    1313524       2400         2400           2400.0    36     15.96   
1076863    1277178      10000        10000          10000.0    36     13.49   
1075358    1311748       3000         3000           3000.0    60     12.69   

         installment grade sub_grade                 emp_title emp_length  \
id                                                                          
1077501       162.87     B        B2                       NaN  10+ years   
1077430        59.83     C        C4                     Ryder   < 1 year   
1077175        84.33     C        C5                       NaN  10+ years   
1076863       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
1075358        67.79     B        B5  University Medical Group     1 year   

        home_ownership  annual_inc verification_status issue_d  loan_status  \
id                                                                            
1077501           RENT     24000.0            Verified  Dec-11   Fully Paid   
1077430           RENT     30000.0     Source Verified  Dec-11  Charged Off   
1077175           RENT     12252.0        Not Verified  Dec-11   Fully Paid   
1076863           RENT     49200.0     Source Verified  Dec-11   Fully Paid   
1075358           RENT     80000.0     Source Verified  Dec-11      Current   

                purpose                 title zip_code addr_state    dti  \
id                                                                         
1077501     credit_card              Computer    860xx         AZ  27.65   
1077430             car                  bike    309xx         GA   1.00   
1077175  small_business  real estate business    606xx         IL   8.72   
1076863           other              personel    917xx         CA  20.00   
1075358           other              Personal    972xx         OR  17.94   

         delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  pub_rec  \
id                                                                         
1077501            0           Jan-85               1         3        0   
1077430            0           Apr-99               5         3        0   
1077175            0           Nov-01               2         2        0   
1076863            0           Feb-96               1        10        0   
1075358            0           Jan-96               0        15        0   

         revol_bal  revol_util  total_acc  out_prncp  out_prncp_inv  \
id                                                                    
1077501      13648        84.0          9       0.00           0.00   
1077430       1687         9.0          4       0.00           0.00   
1077175       2956        98.0         10       0.00           0.00   
1076863       5598        21.0         37       0.00           0.00   
1075358      27783        54.0         38     524.06         524.06   

          total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
id                                                                       
1077501   5863.155187          5833.84          5000.00         863.16   
1077430   1008.710000          1008.71           456.46         435.17   
1077175   3005.666844          3005.67          2400.00         605.67   
1076863  12231.890000         12231.89         10000.00        2214.92   
1075358   3513.330000          3513.33          2475.94        1037.39   

         total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  \
id                                                                              
1077501                0.00        0.00                     0.00       Jan-15   
1077430                0.00      11

In [339]:
# Lets do the data imputation for emp_title and emp_length columns
#Fill the values with mode of the data, as it makes more sense to do this.
df.emp_title = df.emp_title.fillna(df.emp_title.mode()[0])
df.emp_length = df.emp_length.fillna(df.emp_length.mode()[0])
df.head()

member_id  loan_amnt  funded_amnt  funded_amnt_inv  term  int_rate  \
id                                                                            
1077501    1296599       5000         5000           4975.0    36     10.65   
1077430    1314167       2500         2500           2500.0    60     15.27   
1077175    1313524       2400         2400           2400.0    36     15.96   
1076863    1277178      10000        10000          10000.0    36     13.49   
1075358    1311748       3000         3000           3000.0    60     12.69   

         installment grade sub_grade                 emp_title emp_length  \
id                                                                          
1077501       162.87     B        B2                   US Army  10+ years   
1077430        59.83     C        C4                     Ryder   < 1 year   
1077175        84.33     C        C5                   US Army  10+ years   
1076863       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
1075358        67.79     B        B5  University Medical Group     1 year   

        home_ownership  annual_inc verification_status issue_d  loan_status  \
id                                                                            
1077501           RENT     24000.0            Verified  Dec-11   Fully Paid   
1077430           RENT     30000.0     Source Verified  Dec-11  Charged Off   
1077175           RENT     12252.0        Not Verified  Dec-11   Fully Paid   
1076863           RENT     49200.0     Source Verified  Dec-11   Fully Paid   
1075358           RENT     80000.0     Source Verified  Dec-11      Current   

                purpose                 title zip_code addr_state    dti  \
id                                                                         
1077501     credit_card              Computer    860xx         AZ  27.65   
1077430             car                  bike    309xx         GA   1.00   
1077175  small_business  real estate business    606xx         IL   8.72   
1076863           other              personel    917xx         CA  20.00   
1075358           other              Personal    972xx         OR  17.94   

         delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  pub_rec  \
id                                                                         
1077501            0           Jan-85               1         3        0   
1077430            0           Apr-99               5         3        0   
1077175            0           Nov-01               2         2        0   
1076863            0           Feb-96               1        10        0   
1075358            0           Jan-96               0        15        0   

         revol_bal  revol_util  total_acc  out_prncp  out_prncp_inv  \
id                                                                    
1077501      13648        84.0          9       0.00           0.00   
1077430       1687         9.0          4       0.00           0.00   
1077175       2956        98.0         10       0.00           0.00   
1076863       5598        21.0         37       0.00           0.00   
1075358      27783        54.0         38     524.06         524.06   

          total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
id                                                                       
1077501   5863.155187          5833.84          5000.00         863.16   
1077430   1008.710000          1008.71           456.46         435.17   
1077175   3005.666844          3005.67          2400.00         605.67   
1076863  12231.890000         12231.89         10000.00        2214.92   
1075358   3513.330000          3513.33          2475.94        1037.39   

         total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  \
id                                                                              
1077501                0.00        0.00                     0.00       Jan-15   
1077430                0.00      11

In [340]:
#We need dates so that we can do some arithmetic later.
getDate = lambda x : datetime.strptime(x.str.strip(), '%b-%y').date()
df['issue_d'] = df['issue_d'].apply(getDate)
df['issue_year'] = df['issue_d'].apply(lambda x: x.year)
df['issue_month'] = df['issue_d'].apply(lambda x: x.month)

#print (df[['issue_d_tmp', 'issue_year', 'issue_month']])

def MMM_YY_2_Date(x):
    y = datetime.strptime(x, '%b-%y')
    _2019Date = datetime(2019, 1, 1)
    if y > _2019Date:
        y = y - relativedelta(years = 100)
    return y.date()

df['earliest_cr_line'] = df['earliest_cr_line'].apply(MMM_YY_2_Date)
df['earl_cr_line_month'] = df['earliest_cr_line'].apply(lambda x: x.strftime('%b'))
df['earl_cr_line_year'] = df['earliest_cr_line'].apply(lambda x: x.year)

df['credit_history'] = ((df['issue_d'] - df['earliest_cr_line']) / np.timedelta64(1, 'Y')).round()

In [341]:
# Lets standardize the empl_length column
# Less than 1 year be considered 0
def currentEmploymentLength(y):
    if y == None:
        return "invalid"
    y = y.strip()
    if (y.find("< 1") != -1):
        return 0
    result = re.findall("\d+", y)
    if result == None:
        return "invalid"
    return int(result[0])

#Check unique values to know pattern
#print(df.emp_length.value_counts())
df.emp_length = df.emp_length.apply(currentEmploymentLength)
#Validate if the values are as expected
#print(df.emp_length.value_counts())


In [342]:
df['last_credit_pull_month'] = df.last_credit_pull_d.str.split('-').str[0].str.upper()
df['last_credit_pull_Year'] = df.last_credit_pull_d.str.split('-').str[1]

def getLastPaymentDate(x):
    x = str(x).strip()
    ret = datetime(2000, 1, 1)
    if not x:
        return ret
    elif x.upper() == str(float('nan')).upper():
        return ret
    else:    
        ret = datetime.strptime(str(x), '%b-%y').date()
    return ret
    
df['last_pymnt_d_tmp'] = df['last_pymnt_d'].apply(getLastPaymentDate)
#df['last_pymnt_d_tmp'].head()

#Can we get the month and year string?
df['last_pymnt_month'] = df['last_pymnt_d_tmp'].apply(lambda x: x.strftime('%b'))
df['last_pymnt_year'] = df['last_pymnt_d_tmp'].apply(lambda x: x.year)

#df[['last_pymnt_d_tmp', 'last_pymnt_month', 'last_pymnt_year']].head()

In [379]:
#Convert to lower case
df['loan_status'] = df['loan_status'].apply(str.lower)
df['verification_status'] = df['verification_status'].apply(str.lower)
df['purpose'] = df['purpose'].apply(str.lower)
df['title'] = df['title'].astype(str).apply(str.lower)

#EMI as a percent of monthly income
df['emi_income_percent'] = (df['installment'] / (df['annual_inc'] /12)) * 100


In [380]:
#Create some filters 
paidOff = df['loan_status'] == 'fully paid'
current = df['loan_status'] == 'current'
chargedOff = df['loan_status'] == 'charged off'

dfPaidOff = df.loc[paidOff]
dfCurrent = df.loc[current]
dfChargedOff = df.loc[chargedOff]

print (dfPaidOff['loan_status'].size)
print (dfCurrent['loan_status'].size)
print (dfChargedOff['loan_status'].size)

df.describe()

32950
1140
5627


member_id     loan_amnt   funded_amnt  funded_amnt_inv  \
count  3.971700e+04  39717.000000  39717.000000     39717.000000   
mean   8.504636e+05  11219.443815  10947.713196     10397.448868   
std    2.656783e+05   7456.670694   7187.238670      7128.450439   
min    7.069900e+04    500.000000    500.000000         0.000000   
25%    6.667800e+05   5500.000000   5400.000000      5000.000000   
50%    8.508120e+05  10000.000000   9600.000000      8975.000000   
75%    1.047339e+06  15000.000000  15000.000000     14400.000000   
max    1.314167e+06  35000.000000  35000.000000     35000.000000   

               term      int_rate   installment    emp_length    annual_inc  \
count  39717.000000  39717.000000  39717.000000  39717.000000  3.971700e+04   
mean      42.418007     12.021177    324.561922      5.109500  6.896893e+04   
std       10.622815      3.724825    208.874874      3.607467  6.379377e+04   
min       36.000000      5.420000     15.690000      0.000000  4.000000e+03   
25%       36.000000      9.250000    167.020000      2.000000  4.040400e+04   
50%       36.000000     11.860000    280.220000      5.000000  5.900000e+04   
75%       60.000000     14.590000    430.780000     10.000000  8.230000e+04   
max       60.000000     24.590000   1305.190000     10.000000  6.000000e+06   

                dti   delinq_2yrs  inq_last_6mths      open_acc       pub_rec  \
count  39717.000000  39717.000000    39717.000000  39717.000000  39717.000000   
mean      13.315130      0.146512        0.869200      9.294408      0.055065   
std        6.678594      0.491812        1.070219      4.400282      0.237200   
min        0.000000      0.000000        0.000000      2.000000      0.000000   
25%        8.170000      0.000000        0.000000      6.000000      0.000000   
50%       13.400000      0.000000        1.000000      9.000000      0.000000   
75%       18.600000      0.000000        1.000000     12.000000      0.000000   
max       29.990000     11.000000        8.000000     44.000000      4.000000   

           revol_bal    revol_util     total_acc     out_prncp  out_prncp_inv  \
count   39717.000000  39667.000000  39717.000000  39717.000000   39717.000000   
mean    13382.528086     48.833514     22.088828     51.227887      50.989768   
std     15885.016641     28.339127     11.401709    375.172839     373.824457   
min         0.000000      0.000000      2.000000      0.000000       0.000000   
25%      3703.000000     25.000000     13.000000      0.000000       0.000000   
50%      8850.000000     49.000000     20.000000      0.000000       0.000000   
75%     17058.000000     72.000000     29.000000      0.000000       0.000000   
max    149588.000000    100.000000     90.000000   6311.470000    6307.370000   

        total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
count  39717.000000     39717.000000     39717.000000   39717.000000   
mean   12153.596544     11567.149118      9793.348813    2263.663172   
std     9042.040766      8942.672613      7065.522127    2608.111964   
min        0.000000         0.000000         0.000000       0.000000   
25%     5576.930000      5112.310000      4600.000000     662.180000   
50%     9899.640319      9287.150000      8000.000000    1348.910000   
75%    16534.433040     15798.810000     13653.260000    2833.400000   
max    58563.679930     58563.680000     35000.020000   23563.680000   

       total_rec_late_fee    recoveries  collection_recovery_fee  \
count        39717.000000  39717.000000             39717.000000   
mean             1.363015     95.221624                12.406112   
std              7.289979    688.744771               148.671593   
min              0.000000      0.000000                 0.000000   
25%              0.000000      0.000000                 0.000000   
50%              0.000000      0.000000                 0.000000   
75%              0.000000      0.000000                 0.000000   
max            180.200000  29623

In [83]:
#####COPY FROM HERE#########




member_id  loan_amnt  funded_amnt  funded_amnt_inv  term  int_rate  \
id                                                                            
1077501    1296599       5000         5000           4975.0    36     10.65   
1077430    1314167       2500         2500           2500.0    60     15.27   
1077175    1313524       2400         2400           2400.0    36     15.96   
1076863    1277178      10000        10000          10000.0    36     13.49   
1075358    1311748       3000         3000           3000.0    60     12.69   

         installment grade sub_grade                 emp_title emp_length  \
id                                                                          
1077501       162.87     B        B2                   US Army  10+ years   
1077430        59.83     C        C4                     Ryder   < 1 year   
1077175        84.33     C        C5                   US Army  10+ years   
1076863       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
1075358        67.79     B        B5  University Medical Group     1 year   

        home_ownership  annual_inc verification_status issue_d  loan_status  \
id                                                                            
1077501           RENT     24000.0            Verified  Dec-11   Fully Paid   
1077430           RENT     30000.0     Source Verified  Dec-11  Charged Off   
1077175           RENT     12252.0        Not Verified  Dec-11   Fully Paid   
1076863           RENT     49200.0     Source Verified  Dec-11   Fully Paid   
1075358           RENT     80000.0     Source Verified  Dec-11      Current   

                purpose                 title zip_code addr_state    dti  \
id                                                                         
1077501     credit_card              Computer    860xx         AZ  27.65   
1077430             car                  bike    309xx         GA   1.00   
1077175  small_business  real estate business    606xx         IL   8.72   
1076863           other              personel    917xx         CA  20.00   
1075358           other              Personal    972xx         OR  17.94   

         delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  pub_rec  \
id                                                                         
1077501            0           Jan-85               1         3        0   
1077430            0           Apr-99               5         3        0   
1077175            0           Nov-01               2         2        0   
1076863            0           Feb-96               1        10        0   
1075358            0           Jan-96               0        15        0   

         revol_bal  revol_util  total_acc  out_prncp  out_prncp_inv  \
id                                                                    
1077501      13648        84.0          9       0.00           0.00   
1077430       1687         9.0          4       0.00           0.00   
1077175       2956        98.0         10       0.00           0.00   
1076863       5598        21.0         37       0.00           0.00   
1075358      27783        54.0         38     524.06         524.06   

          total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
id                                                                       
1077501   5863.155187          5833.84          5000.00         863.16   
1077430   1008.710000          1008.71           456.46         435.17   
1077175   3005.666844          3005.67          2400.00         605.67   
1076863  12231.890000         12231.89         10000.00        2214.92   
1075358   3513.330000          3513.33          2475.94        1037.39   

         total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  \
id                                                                              
1077501                0.00        0.00                     0.00       Jan-15   
1077430                0.00      11

In [84]:
# Lets check the details of addr_state
print(df.addr_state.unique())

# lets create Region column depedning on the state column value
# https://www.ducksters.com/geography/us_states/us_geographical_regions.php
west = ['CA', 'OR', 'UT','WA', 'CO', 'NV', 'AK', 'MT', 'HI', 'WY', 'ID']
south_west = ['AZ', 'TX', 'NM', 'OK']
south_east = ['GA', 'NC', 'VA', 'FL', 'KY', 'SC', 'LA', 'AL', 'WV', 'DC', 'AR', 'DE', 'MS', 'TN' ]
mid_west = ['IL', 'MO', 'MN', 'OH', 'WI', 'KS', 'MI', 'SD', 'IA', 'NE', 'IN', 'ND']
north_east = ['CT', 'NY', 'PA', 'NJ', 'RI','MA', 'MD', 'VT', 'NH', 'ME']

df['Region']= np.nan

def map_region(state):
    if state in west:
        return 'west'
    elif state in south_west:
        return 'south_west'
    elif state in south_east:
        return 'south_east'
    elif state in mid_west:
        return 'mid_west'
    elif state in north_east:
        return 'north_east'
df['Region'] = df['addr_state'].apply(map_region)
df.head()

['AZ' 'GA' 'IL' 'CA' 'OR' 'NC' 'TX' 'VA' 'MO' 'CT' 'UT' 'FL' 'NY' 'PA'
 'MN' 'NJ' 'KY' 'OH' 'SC' 'RI' 'LA' 'MA' 'WA' 'WI' 'AL' 'CO' 'KS' 'NV'
 'AK' 'MD' 'WV' 'VT' 'MI' 'DC' 'SD' 'NH' 'AR' 'NM' 'MT' 'HI' 'WY' 'OK'
 'DE' 'MS' 'TN' 'IA' 'NE' 'ID' 'IN' 'ME']


member_id  loan_amnt  funded_amnt  funded_amnt_inv  term  int_rate  \
id                                                                            
1077501    1296599       5000         5000           4975.0    36     10.65   
1077430    1314167       2500         2500           2500.0    60     15.27   
1077175    1313524       2400         2400           2400.0    36     15.96   
1076863    1277178      10000        10000          10000.0    36     13.49   
1075358    1311748       3000         3000           3000.0    60     12.69   

         installment grade sub_grade                 emp_title emp_length  \
id                                                                          
1077501       162.87     B        B2                   US Army  10+ years   
1077430        59.83     C        C4                     Ryder   < 1 year   
1077175        84.33     C        C5                   US Army  10+ years   
1076863       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
1075358        67.79     B        B5  University Medical Group     1 year   

        home_ownership  annual_inc verification_status issue_d  loan_status  \
id                                                                            
1077501           RENT     24000.0            Verified  Dec-11   Fully Paid   
1077430           RENT     30000.0     Source Verified  Dec-11  Charged Off   
1077175           RENT     12252.0        Not Verified  Dec-11   Fully Paid   
1076863           RENT     49200.0     Source Verified  Dec-11   Fully Paid   
1075358           RENT     80000.0     Source Verified  Dec-11      Current   

                purpose                 title zip_code addr_state    dti  \
id                                                                         
1077501     credit_card              Computer    860xx         AZ  27.65   
1077430             car                  bike    309xx         GA   1.00   
1077175  small_business  real estate business    606xx         IL   8.72   
1076863           other              personel    917xx         CA  20.00   
1075358           other              Personal    972xx         OR  17.94   

         delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  pub_rec  \
id                                                                         
1077501            0           Jan-85               1         3        0   
1077430            0           Apr-99               5         3        0   
1077175            0           Nov-01               2         2        0   
1076863            0           Feb-96               1        10        0   
1075358            0           Jan-96               0        15        0   

         revol_bal  revol_util  total_acc  out_prncp  out_prncp_inv  \
id                                                                    
1077501      13648        84.0          9       0.00           0.00   
1077430       1687         9.0          4       0.00           0.00   
1077175       2956        98.0         10       0.00           0.00   
1076863       5598        21.0         37       0.00           0.00   
1075358      27783        54.0         38     524.06         524.06   

          total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
id                                                                       
1077501   5863.155187          5833.84          5000.00         863.16   
1077430   1008.710000          1008.71           456.46         435.17   
1077175   3005.666844          3005.67          2400.00         605.67   
1076863  12231.890000         12231.89         10000.00        2214.92   
1075358   3513.330000          3513.33          2475.94        1037.39   

         total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  \
id                                                                              
1077501                0.00        0.00                     0.00       Jan-15   
1077430                0.00      11

member_id  loan_amnt  funded_amnt  funded_amnt_inv  term  int_rate  \
id                                                                            
1077501    1296599       5000         5000           4975.0    36     10.65   
1077430    1314167       2500         2500           2500.0    60     15.27   
1077175    1313524       2400         2400           2400.0    36     15.96   
1076863    1277178      10000        10000          10000.0    36     13.49   
1075358    1311748       3000         3000           3000.0    60     12.69   

         installment grade sub_grade                 emp_title  emp_length  \
id                                                                           
1077501       162.87     B        B2                   US Army          10   
1077430        59.83     C        C4                     Ryder           1   
1077175        84.33     C        C5                   US Army          10   
1076863       339.31     C        C1       AIR RESOURCES BOARD          10   
1075358        67.79     B        B5  University Medical Group           1   

        home_ownership  annual_inc verification_status issue_d  loan_status  \
id                                                                            
1077501           RENT     24000.0            Verified  Dec-11   Fully Paid   
1077430           RENT     30000.0     Source Verified  Dec-11  Charged Off   
1077175           RENT     12252.0        Not Verified  Dec-11   Fully Paid   
1076863           RENT     49200.0     Source Verified  Dec-11   Fully Paid   
1075358           RENT     80000.0     Source Verified  Dec-11      Current   

                purpose                 title zip_code addr_state    dti  \
id                                                                         
1077501     credit_card              Computer    860xx         AZ  27.65   
1077430             car                  bike    309xx         GA   1.00   
1077175  small_business  real estate business    606xx         IL   8.72   
1076863           other              personel    917xx         CA  20.00   
1075358           other              Personal    972xx         OR  17.94   

         delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  pub_rec  \
id                                                                         
1077501            0           Jan-85               1         3        0   
1077430            0           Apr-99               5         3        0   
1077175            0           Nov-01               2         2        0   
1076863            0           Feb-96               1        10        0   
1075358            0           Jan-96               0        15        0   

         revol_bal  revol_util  total_acc  out_prncp  out_prncp_inv  \
id                                                                    
1077501      13648        84.0          9       0.00           0.00   
1077430       1687         9.0          4       0.00           0.00   
1077175       2956        98.0         10       0.00           0.00   
1076863       5598        21.0         37       0.00           0.00   
1075358      27783        54.0         38     524.06         524.06   

          total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
id                                                                       
1077501   5863.155187          5833.84          5000.00         863.16   
1077430   1008.710000          1008.71           456.46         435.17   
1077175   3005.666844          3005.67          2400.00         605.67   
1076863  12231.890000         12231.89         10000.00        2214.92   
1075358   3513.330000          3513.33          2475.94        1037.39   

         total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  \
id                                                                              
1077501                0.00        0.00                     0.00       Jan-15   
1077430                0.00 

## Insights and conclusions on loan defaulters (charged-off)

### Note: Siddakka, every cell in insight should focus on one insight / conclusion. There might be a preceeding step related to a data wrangling cleaning, deriving step. All the supporting analysis, can be moved to the end. 

### Every insight cell will have data local to itself. By doing this the code's readability is very good. The dimensions with which we are analysing is too many and hence the names will get lost. Imagine we have a dfXXX_YYY_ZZZ created in the data clean up in cell 10 and we use it in cell 55. The reviewer need not go to cell 10 to see that data.

### Step 1. Load the required libraries, opend the loan.csv and cleanup data
##### Note: If we are dropping columns, imputing, we need to explain why we are doing. We can refer to a step in future in "EDA experiments supporting insights ad conclusions". We will have a brief EDA step here in this section, but not too graphical that distracts the reader / reviewer.

### Step 2. Derive data for simplicity and ease of analysis.
##### Note: Standardization and data formating of columns ---> goes here. We can just retain charged-off customers, because thats what we need to focus upon. 
##### Add functions in this section that will be used later.
##### Your region categorization function should be defined here

### Step 3. Do preliminary EDA and understand data.
### Lets do the Loan similarity distribution
- What amount of loan has been issued
- What is frequency distribution loan over a period
## Good Loan vs Bad Loans
**Lets see the no. of Bad loans in Dataframe to find out the risk of load falling into default category.**

**From the above graph its clear that the loans that are charged off are the bad loans. Aound 14.2 % of loans are bad loans**

**From the above graph we can conclude that most of loans were issued to grade A and B**

## Lets do the Analysi of Bad loan vs different variables

### Lets see the distribution of the annual income of the employees

#### From the above graphs we can conclude the below points
- The amount of loan issued lies between 5000 to 15000 USD
- The annual income of the most applicants is <1000 k USD

**From the above graph we can conclude that Amount of the loan issued is not dependent on the Loan status
  
**From the above graph we can conclude that the Grade A and B have received the loan with minimum interest rate and Grade G has received loan with maximum interest rate**

**From the above graph we can conclude that the interest rate of charged off loans is greater then the fully paid**

**Lets check the relation between the Loan purpose and the Loan status**
### From the above graph we can conclude below points:
1. Loans Issued for the perpose of Debt Consolidation are being paid off
2. Loans Issued for the perpose of Debt Consolidation are being Charged off

### Lets see if there is any impact of terms i.e. the loan terms in months on the loan status
### Lets check the relation between home ownership and the Loan status

##### Do Top4 region, Top4 state, Top4 reason, Top4 month and Terms and explain why you need that and not the entire data set.

## All the cells of Siddakka till cell 50 go here. We need to change the colors similar to cell 42. its very sober and smooth.

### Step 4. We are retaining only charged-off borrowers for analysis. Why?

##### Why? Explain the reasons.

### Step 5. Term and Interest rate analysis

#### OBSERVATION - For those with 36 months term, about (800+650) approx 1500 of them dropped out when interest rate was 12 - 14 percent. From 60 months term, about 900 of them dropped out when interest rate was 15 - 17 percent. Observe that the defaulter count is low for 36 months as interest rate increases. 
#### CONCLUSION - So for 36 months, the interest rate has to be higher.
#### SUPPORTING EDA - Refer Section number XX for extensive EDA that helped us arrive at this conclusion.

### Step 6. Term with credit_history (derived from earliest credit line and loan issued year). Note that the defaulters drastically reduce for both terms 36 and 60 as the credit_history (age of the person) increases. Its risky to give loans to customers with credit history between 5 to 15 years.

#### OBSERVATION - xxxxxx 
#### CONCLUSION - xxxx
#### SUPPORTING EDA - Refer Section number XX for extensive EDA that helped us arrive at this conclusion.

### Step 7. Analyse MONTH loan issued, PURPOSE of loan and STATE
#### OBSERVATION - xxxxxx 
#### CONCLUSION - xxxx
#### SUPPORTING EDA - Refer Section number XX for extensive EDA that helped us arrive at this conclusion.

### Step 8. From defaulter data, we got TOP 4 - Months, States and Purpose. We build a pivot table to categorize the purpose of loan for a month for a state. Here we can obviously see that most of defaulters take loans under the category debt consolidation, in December across all 4 states.

#### OBSERVATION - xxxxxx 
#### CONCLUSION - xxxx
#### SUPPORTING EDA - Refer Section number XX for extensive EDA that helped us arrive at this conclusion.

### Step 9. In November and December, in California, when a consumer takes loan under debt consolidation does credit history length have any impact. YES THIS CHART INDICATES credit history of 10 - 15 years definitely is a risk

#### OBSERVATION - xxxxxx 
#### CONCLUSION - xxxx
#### SUPPORTING EDA - Refer Section number XX for extensive EDA that helped us arrive at this conclusion.

### Step 10. Now that we know in last quater, especially December, customers take debt consolidation loan and default. Is employment history affecting their default status. YES, since 10 means greater than 10 years, we can ignore that. There is a pattern that customers with less than 1 year experience, default more and it reduces as employment length increases.
##### When a customer comes in Nov, Dec with debt consolidation as the reason and less than 1 year of employment length in current job, its risky.

#### OBSERVATION - xxxxxx 
#### CONCLUSION - xxxx
#### SUPPORTING EDA - Refer Section number XX for extensive EDA that helped us arrive at this conclusion.

### Step 11. Now that we know in last quater, especially December, customers take debt consolidation loan and default. Does home ownership affect defaulting nature. Yes, those who rent default significantly higher than mortagers and mortgagers default higher than home owners. Hence, if a customer in rental home applies for loan with purpose debt consolidation in the months November - December, its risky.
#### OBSERVATION - xxxxxx 
#### CONCLUSION - xxxx
#### SUPPORTING EDA - Refer Section number XX for extensive EDA that helped us arrive at this conclusion.

### Step 9 in original file - IGNORE THIS or move it to supplement section.  Now that we know in last quater, especially December, customers take debt consolidation loan and default. How does revol_util affect the defaulting tendency. NO distinct pattern in this - READ BELOW section 14. WE have proved that revol_util has a BIG IMPACT.

#### OBSERVATION - xxxxxx 
#### CONCLUSION - xxxx
#### SUPPORTING EDA - Refer Section number XX for extensive EDA that helped us arrive at this conclusion.



### Step 12. Now that we know in last quater, especially December, customers take debt consolidation loan and default. Income verficiation has any relationship? YES - If income source verification is done, the risk is LOWER
#### OBSERVATION - xxxxxx 
#### CONCLUSION - xxxx
#### SUPPORTING EDA - Refer Section number XX for extensive EDA that helped us arrive at this conclusion.

### Step 13. Now that we know in last quater, especially December, customers take debt consolidation loan and default.
### How many months after these customers take loan they default?
### How much money was recovered in total from them?
#### OBSERVATION - xxxxxx 
#### CONCLUSION - xxxx

#### Conclusion: The below chart is for top 4 months, top 4 purpose, top 4 states defaulters, who are in rented homes. We index it based on the length of employment. It clearly shows that, customers with less than 1 year of current employment experience default the most.

#### SUPPORTING EDA - Refer Section number XX for extensive EDA that helped us arrive at this conclusion.

### Step 14. For the months Nov, Dec, in the state of California, those customers who take loan with reason as debt consolidation, we want to know what is the recovery rate. Recovery rate is the amount recovered as a percent of funded amount. The amount we refer is  principal + interest + other charges, which is total_pymnt. YES! There is a pattern that the MEDIAN recovery rate is around 40% for those living in rented house and highest with those in their own home.

### Also for CA state the median recovery rate is only 53%. If you replace index to be 'addr_state', you can observe this. This indicates that customers in rented house bring down the median a lot.
#### OBSERVATION - xxxxxx 
#### CONCLUSION - xxxx

#### SUPPORTING EDA - Refer Section number XX for extensive EDA that helped us arrive at this conclusion.

### Step 15.  For the months Nov, Dec, in the state of California, those customers who take loan with reason as debt consolidation, we want to know the DTI relationship. A very powerful pattern emerges here. 
##### The Median DTI is high and close to home owners. Considering the previous chart, since recovery is lowest amongst rental home owners, and DTI is higer than mortagers, its risky to lend money to rental owners.
##### Also when we seggregate the DTI w.r.t credit history, we see that its in the order of
##### 0 - 15 for Home owners
##### 0 - 20 for Mortgagers and
##### 0 - 40 for Rental folks - Wide spread.
##### This is a powerful observation that rental home borrowers have significantly higher and broader span of DTI, which implies that rental owners have more debt than income.
#### OBSERVATION - xxxxxx 
#### CONCLUSION - xxxx

#### SUPPORTING EDA - Refer Section number XX for extensive EDA that helped us arrive at this conclusion.

### Step 16. For the months Nov, Dec, in the state of California, those customers who take loan with reason as debt consolidation, we want to know the revolving utilization impact on defaulters. A very powerful pattern emerges here. 

##### For rental customers, revolving utilization is highest, and specifically, it reduces as they spend more number of years with the same employer. So many reasons for calling rental clients risky.

## EDA experiments supporting insights and conclusions